packages/@lwc/engine-core/src/patches/detect-synthetic-cross-root-aria.ts

```diff
@@ -0,0 +1,183 @@
+                }
+            }
+        }
+    }
+}
+
+let enabled = false;
+
+// We want to avoid patching globals whenever possible, so this should be tree-shaken out in prod-mode and if
+// reporting is not enabled. It should also only run once
+function enableDetection() {
+    if (enabled) {
+        return; // don't double-apply the patches
+    }
+    enabled = true;
+
+    const { setAttribute } = Element.prototype;
+
+    // Detect calling `setAttribute` to set an idref or an id
+    assign(Element.prototype, {
+        setAttribute(this: Element, attrName: string, attrValue: any) {
+            setAttribute.call(this, attrName, attrValue);
+            if (attrName === 'id' || ID_REFERENCING_ATTRIBUTES_SET.has(attrName)) {
+                detectSyntheticCrossRootAria(this, attrName, attrValue);
+            }
+        },
+    } as Pick<Element, 'setAttribute'>);
+
+    // Detect `elm.id = 'foo'`
+    const idDescriptor = getOwnPropertyDescriptor(Element.prototype, 'id');
+    if (!isUndefined(idDescriptor)) {
+        const { get, set } = idDescriptor;
+        // These should always be a getter and a setter, but if someone is monkeying with the global descriptor, ignore it
+        if (isFunction(get) && isFunction(set)) {
+            defineProperty(Element.prototype, 'id', {
+                get() {
+                    return get.call(this);
+                },
+                set(value: any) {
+                    set.call(this, value);
+                    detectSyntheticCrossRootAria(this, 'id', value);
+                },
+                // On the default descriptor for 'id', enumerable and configurable are true
+                enumerable: true,
+                configurable: true,
+            });
+        }
+    }
+}
+
+// Our detection logic relies on some modern browser features. We can just skip reporting the data
+// for unsupported browsers
+function supportsCssEscape() {
+    return typeof CSS !== 'undefined' && isFunction(CSS.escape);
+}
+
+// If this page is not using synthetic shadow, then we don't need to install detection. Note
+// that we are assuming synthetic shadow is loaded before LWC.
+function isSyntheticShadowLoaded() {
+    // We should probably be calling `renderer.isSyntheticShadowDefined`, but 1) we don't have access to the renderer,
+    // and 2) this code needs to run in @lwc/engine-core, so it can access `logWarn()` and `report()`.
+    return hasOwnProperty.call(Element.prototype, KEY__SHADOW_TOKEN);
+}
+
+// Detecting cross-root ARIA in synthetic shadow only makes sense for the browser
+if (process.env.IS_BROWSER && supportsCssEscape() && isSyntheticShadowLoaded()) {
+    // Always run detection in dev mode, so we can at least print to the console
+    if (process.env.NODE_ENV !== 'production') {
+        enableDetection();
+    } else {
+        // In prod mode, only enable detection if reporting is enabled
+        onReportingEnabled(enableDetection);
```

Comment:
If the reporting API is unused (i.e. `__unstable__ReportingControl` is not imported), then all of this code should be tree-shaken out. I tested manually with Terser and it worked.